File: word-search/word-search.cpp
class Solution {
public:

    bool isSafe(int i,int j,int n,int m) {
        if(i>=0 && i<n && j>=0 && j<m) return true;
        else return false;
    }

    bool dfs(vector<vector<char>>& board,string word,vector<vector<bool>>& vis,int n,int i,int j) {
        if(n==word.length()) return true;
        vis[i][j]=true;

        if(isSafe(i,j+1,board.size(),board[0].size()) && !vis[i][j+1] && word[n]==board[i][j+1]) 
        if(dfs(board,word,vis,n+1,i,j+1)) return true;
        if(isSafe(i,j-1,board.size(),board[0].size()) && !vis[i][j-1] && word[n]==board[i][j-1]) 
        if(dfs(board,word,vis,n+1,i,j-1)) return true;
        if(isSafe(i+1,j,board.size(),board[0].size()) && !vis[i+1][j] && word[n]==board[i+1][j]) 
        if(dfs(board,word,vis,n+1,i+1,j)) return true;
        if(isSafe(i-1,j,board.size(),board[0].size()) && !vis[i-1][j] && word[n]==board[i-1][j]) 
        if(dfs(board,word,vis,n+1,i-1,j)) return true;

        vis[i][j]=false;
        return false;
    }

    bool exist(vector<vector<char>>& board, string word) {
        vector<vector<bool>> vis(board.size(),vector<bool>(board[0].size(),false));
        for(int i=0;i<board.size();i++)
        {
            for(int j=0;j<board[0].size();j++)
            {
                if(board[i][j]==word[0] && dfs(board,word,vis,1,i,j)) return true;
            }
        }
        return false;
    }
};

File: 79-word-search/word-search.java
class Solution {

    int[][] coord = {{-1,0},{0,-1},{0,1},{1,0}};

    public boolean dfs(char[][] board,String word,boolean[][] vis,int i,int j,int ind,int m,int n)
    {
        if(ind>=word.length()) return true;
        if(i<0 || i>=m || j<0 || j>=n || vis[i][j]==true || word.charAt(ind)!=board[i][j]) return false;
        vis[i][j]=true;

        for(int in = 0;in<4;in++)
        {
            int x = coord[in][0];
            int y = coord[in][1];
            if(dfs(board,word,vis,i+x,j+y,ind+1,m,n)) return true;
        }
        vis[i][j]=false;
        return false;
    }

    public boolean exist(char[][] board, String word) {
        int m = board.length;
        int n = board[0].length;
        boolean[][] vis= new boolean[m][n];
        for(int i=0;i<m;i++)
        {
            for(int j=0;j<n;j++)
            {
                if(board[i][j]==word.charAt(0))
                {
                    if(dfs(board,word,vis,i,j,0,m,n)==true) return true;
                }
            }
        }
        return false;
    }
}